posts/2022/08-27-speed-up-ci-migrations.md

```diff
@@ -0,0 +1,212 @@
+
+```bash
+> time mix ecto.reset
+
+Executed in  141.59 secs    fish           external
+   usr time    2.85 secs  600.00 micros    2.85 secs
+   sys time    0.57 secs  145.00 micros    0.57 secs
+```
+
+Whereas using the dump takes only 38 seconds (I have overridden reset in my `mix.exs`)
+
+```bash
+> time mix ecto.reset
+
+Executed in   38.27 secs    fish           external
+   usr time    2.50 secs    0.00 micros    2.50 secs
+   sys time    0.38 secs  712.00 micros    0.38 secs
+```
+
+This can be a huge difference when we are talking about this executing on **every single CI run**.
+
+## What is a structure dump and why is it faster than migrating?
+
+A database structure dump is a record of the database structure, including things like the table, columns, indexes and constraints. It is NOT a record of the data (that would be a regular database dump), so you can expect all your tables to be empty of rows when loading a dump.
+
+Instead of iterating through migrations one at a time, and having each run in a transaction, the whole database can be created from one file. This dramatically decreases the time necessary to create a database from scratch!
+
+[Read more about pg_dump](https://www.postgresql.org/docs/current/app-pgdump.html)
+
+## Are there disadvantages to using database structure dumps?
+
+Yes. Once your migrations are transferred into the `dumped_migrations` folder and are no longer actively run on every CI build they will gradually begin to "rot". Helper functions that worked when the migration happened will change in backwards incompatible ways and it will not be discovered until a person, say, moves all those migrations back to the migrations folder and attempts to run them all.
+
+Guess how I figured this out :)
+
+Also, if your team uses migrations to migrate data (which I honestly consider an anti-pattern) this will not work for you and your team. Structure dumps will not insert data and you will need to find another way.
+
+## How we can use dumps alongside migrations
+
+The usual order for generating a database from scratch is:
+
+1. Create a database with `mix ecto.create`
+2. Migrate the database with `mix ecto.migrate`
+
+When using a structure dump file we can add a mix task in between to load the structure.
+
+1. Create a database with `mix ecto.create`
+2. Load the structure dump with `mix ecto.load --force`
+3. Migrate any remaining migrations that are not part of the dump yet with `mix ecto.migrate`
+
+## How to create and update a structure dump
+
+In order to generate an up-to-date structure dump we must make sure that the database has all the migrations run, and after creating the dump we need to move the migrations to another folder so that they are not run again (because they are now part of the dump instead). The process looks like the following:
```

Comment:
In order to generate [...], we must, first of all, make sure that we have run all the migrations for our database. Once we have created the dump, we need to... [...]. This is what the process looks like: